lua/null-ls/builtins/formatting/nix_flake_fmt.lua

```diff
@@ -0,0 +1,145 @@
+local h = require("null-ls.helpers")
+local methods = require("null-ls.methods")
+local log = require("null-ls.logger")
+
+local FORMATTING = methods.internal.FORMATTING
+
+--- Return the command that `nix fmt` would run, or nil if we're not in a
+--- flake.
+---
+--- The formatter must follow treefmt's [formatter
+--- spec](https://github.com/numtide/treefmt/blob/main/docs/formatter-spec.md).
+---
+--- This basically re-implements the "entrypoint discovery" that `nix fmt` does.
+--- So why are we doing this ourselves rather than just invoking `nix fmt`?
+--- Unfortunately, it can take a few moments to evaluate all your nix code to
+--- figure out the formatter entrypoint. It can even be slow enough to exceed
+--- Neovim's default LSP timeout.
+--- By doing this ourselves, we can cache the result.
+---
+---@return string|nil
+local find_nix_fmt = function(params)
+    -- Discovering currentSystem here lets us keep the *next* eval pure.
+    -- We want to keep that part pure as a performance improvement: an impure
+    -- eval that references the flake would copy *all* files (including
+    -- gitignored files!), which can be quite expensive if you've got many GiB
+    -- of artifacts in the directory. This optimization can probably go away
+    -- once the [Lazy trees PR] lands.
+    --
+    -- [Lazy trees PR]: https://github.com/NixOS/nix/pull/6530
+    local cp = vim.system({ "nix", "eval", "--impure", "--expr", "builtins.currentSystem" }):wait()
+    if cp.code ~= 0 then
+        log:warn(string.format("unable discover builtins.currentSystem from nix. stderr: %s", cp.stderr))
+        return nil
+    end
+    local nix_current_system = cp.stdout
+
+    local eval_nix_formatter = [[
+      let
+        currentSystem = ]] .. nix_current_system .. [[;
+        # Various functions vendored from nixpkgs lib (to avoid adding a
+        # dependency on nixpkgs).
+        lib = rec {
+          getOutput = output: pkg:
+            if ! pkg ? outputSpecified || ! pkg.outputSpecified
+            then pkg.${output} or pkg.out or pkg
+            else pkg;
+          getBin = getOutput "bin";
+          # Simplified by removing various type assertions.
+          getExe' = x: y: "${getBin x}/bin/${y}";
+          # getExe is simplified to assume meta.mainProgram is specified.
+          getExe = x: getExe' x x.meta.mainProgram;
+        };
+      in
+      formatterBySystem:
+        if formatterBySystem ? ${currentSystem} then
+          let
+            formatter = formatterBySystem.${currentSystem};
+            drv = formatter.drvPath;
+            bin = lib.getExe formatter;
+          in
+            drv + "\n" + bin + "\n"
+        else
+          ""
+    ]]
+
+    cp = vim.system({ "nix", "eval", ".#formatter", "--raw", "--apply", eval_nix_formatter }, { cwd = params.root })
```

Comment:
needs `--extra-experimental-flags "flakes nix-command"`.

Reply:
Done! FYI, it's `--extra-experimental-features`, not flags ;)